### DISCUSSION TITLE
Error in Task 9: "AttributeError: csr_matrix not found"

### DISCUSSION THREAD

**Discussion by marnixb** [Part-II-Computational-Physics/.github#30] [General]
In task 9, when I run my `SparseHamiltonian(N)` function in a for loop for different N, SciPy throws the following error at N=3:
```
AttributeError                            Traceback (most recent call last)
Cell In[17], line 23
     19     timingCompactTensor.append(resultTensor)
     21 for N in Ns: 
     22     # Time sparse matrix method
---> 23     sparse = SparseHamiltonian(N)
     24     resultSparse = get_ipython().run_line_magic('timeit', '-o spsl.eigsh(sparse, k=N)')
     25     timingCompactSparse.append(result)

Cell In[12], line 10, in SparseHamiltonian(N, h)
      9 def SparseHamiltonian(N, h=1):
---> 10     H = sparse.csr_matrix((2**N, 2**N))
     11     loc_ham = sparse.csr_matrix(LocalHamiltonian(h))
     13     for i in range(N-1):

File /usr/local/lib/python3.9/site-packages/scipy/sparse/_base.py:771, in spmatrix.__getattr__(self, attr)
    769     return self.getnnz()
    770 else:
--> 771     raise AttributeError(attr + " not found")

AttributeError: csr_matrix not found
```

When I run the same `SparseHamiltonian(N)` earlier in the code, it works fine. This might be a SciPy or Jupyter notebook bug.

**Comment by JNHallen** [comment]
That is very strange. I am not sure exactly what causes this. Did you try restarting the kernel and running the code again?

**Reply by marnixb** [reply]
I have tried restarting the kernel and using different python versions, but nothing seems to work. I have now pushed everything, so feel free to run my code yourself.

**Reply by JNHallen** [reply]
Your code works fine on my computer, so I am really not sure what the problem might be here.

**Reply by AustenLamacraft** [reply]
It looks like the problem is that you have a name collision between the library `scipy.sparse` that is imported as `sparse` and the variable `sparse` that is the Hamiltonian. When you do `sparse.csr_matrix` you think you are calling `scipy.sparse.csr_matrix` whereas in fact you are looking for the nonexistent attribute `csr_matrix` of an object of the class `csr_matrix`

**Reply by marnixb** [reply]
Ah yes, I didn't see that one. It works now, thank you!
